# Data wrangling

In this notebook, we will observe the way we treated our dataset to be able to make the most of it. We downloaded it from Wharton's website which we have access to thank to Pr. Fahlenbrach. Unfortunately, all the data came in a single column and with lots of `NaN` values. We used `pandas` library to convert our csv file into a dataframe, reshape the dataset and remove the `NaN` values. We then reconverted it into a csv file. 

_ Note that the original file was 1.83GB and the output is now 262MB. _

In [1]:
import numpy as np
import pandas as pd

In [2]:
path = 'data.csv'

In [3]:
csv_file_data = pd.read_csv(path)

Here is a little insight of the data's shape before wrangling, one can observe that the daily returns of the 21661 companies come stacked in a single column. Our aim is to unstack the data obtain a table with rows corresponding to days over time and columns corresponding to the companies. 

In [4]:
csv_file_data.head()

PERMNO      date TICKER           COMNAM  OPENPRC
0   10001  19970102   EWST  ENERGY WEST INC    8.625
1   10001  19970103   EWST  ENERGY WEST INC    8.625
2   10001  19970106   EWST  ENERGY WEST INC    8.625
3   10001  19970107   EWST  ENERGY WEST INC    8.125
4   10001  19970108   EWST  ENERGY WEST INC    8.125

In [5]:
data_df = csv_file_data.drop('PERMNO', axis = 1).groupby(['date', 'COMNAM']).sum().unstack()
data_df.head(20)

OPENPRC                                        \
COMNAM   012 SMILE COMMUNICATIONS LTD 1 800 ATTORNEY INC 1 800 CONTACTS INC   
date                                                                          
19970102                          NaN                NaN                NaN   
19970103                          NaN                NaN                NaN   
19970106                          NaN                NaN                NaN   
19970107                          NaN                NaN                NaN   
19970108                          NaN                NaN                NaN   
19970109                          NaN                NaN                NaN   
19970110                          NaN                NaN                NaN   
19970113                          NaN                NaN                NaN   
19970114                          NaN                NaN                NaN   
19970115                          NaN                NaN                NaN   
19970116                          NaN                NaN                NaN   
19970117                          NaN                NaN                NaN   
19970120                          NaN                NaN                NaN   
19970121                          NaN                NaN                NaN   
19970122                          NaN                NaN                NaN   
19970123                          NaN                NaN                NaN   
19970124                          NaN                NaN                NaN   
19970127                          NaN                NaN                NaN   
19970128                          NaN                NaN                NaN   
19970129                          NaN                NaN                NaN   

                                                  \
COMNAM   1 800 FLOWERS COM INC 1347 CAPITAL CORP   
date                                               
19970102                   NaN               NaN   
19970103                   NaN               NaN   
19970106                   NaN               NaN   
19970107                   NaN               NaN   
19970108                   NaN               NaN   
19970109                   NaN               NaN   
19970110                   NaN               NaN   
19970113                   NaN               NaN   
19970114                   NaN               NaN   
19970115                   NaN               NaN   
19970116                   NaN               NaN   
19970117                   NaN               NaN   
19970120                   NaN               NaN   
19970121                   NaN               NaN   
19970122                   NaN               NaN   
19970123                   NaN               NaN   
19970124                   NaN               NaN   
19970127                   NaN               NaN   
19970128                   NaN               NaN   
19970129                   NaN               NaN   

                                                                            \
COMNAM   1347 PROPERTY INSURANCE HLDS INC 1838 BOND DEBENTURE TRADING FUND   
date                                                                         
19970102                              NaN                           19.750   
19970103                              NaN                           19.875   
19970106                              NaN                           19.875   
19970107                              NaN                           19.875   
19970108                              NaN                           19.750   
19970109                              NaN                           19.875   
19970110                              NaN                           19.750   
19970113                              NaN                           19.875   
19970114                              NaN                           19.750   
19970115                              NaN                           19.750   
1

In [6]:
data_df.shape

(5046, 21661)

Once the DataFrame has been reshaped, we have a _nicer_ overview of our dataset's elements and can observe a large amount of `NaN` values. We will not drop all of them because that would imply too much loss of data but we will drop the columns where no return has been detected. Moreover, in order to avoid divergence in our further computations (project 2), we will also handle the infinite values encountered by replacing them by `NaN`.

In [7]:
data_df_clean = data_df.replace([np.inf, -np.inf], np.nan).dropna(axis=1,how='all')
data_df_clean.head(20)

OPENPRC                                        \
COMNAM   012 SMILE COMMUNICATIONS LTD 1 800 ATTORNEY INC 1 800 CONTACTS INC   
date                                                                          
19970102                          NaN                NaN                NaN   
19970103                          NaN                NaN                NaN   
19970106                          NaN                NaN                NaN   
19970107                          NaN                NaN                NaN   
19970108                          NaN                NaN                NaN   
19970109                          NaN                NaN                NaN   
19970110                          NaN                NaN                NaN   
19970113                          NaN                NaN                NaN   
19970114                          NaN                NaN                NaN   
19970115                          NaN                NaN                NaN   
19970116                          NaN                NaN                NaN   
19970117                          NaN                NaN                NaN   
19970120                          NaN                NaN                NaN   
19970121                          NaN                NaN                NaN   
19970122                          NaN                NaN                NaN   
19970123                          NaN                NaN                NaN   
19970124                          NaN                NaN                NaN   
19970127                          NaN                NaN                NaN   
19970128                          NaN                NaN                NaN   
19970129                          NaN                NaN                NaN   

                                                  \
COMNAM   1 800 FLOWERS COM INC 1347 CAPITAL CORP   
date                                               
19970102                   NaN               NaN   
19970103                   NaN               NaN   
19970106                   NaN               NaN   
19970107                   NaN               NaN   
19970108                   NaN               NaN   
19970109                   NaN               NaN   
19970110                   NaN               NaN   
19970113                   NaN               NaN   
19970114                   NaN               NaN   
19970115                   NaN               NaN   
19970116                   NaN               NaN   
19970117                   NaN               NaN   
19970120                   NaN               NaN   
19970121                   NaN               NaN   
19970122                   NaN               NaN   
19970123                   NaN               NaN   
19970124                   NaN               NaN   
19970127                   NaN               NaN   
19970128                   NaN               NaN   
19970129                   NaN               NaN   

                                                                            \
COMNAM   1347 PROPERTY INSURANCE HLDS INC 1838 BOND DEBENTURE TRADING FUND   
date                                                                         
19970102                              NaN                           19.750   
19970103                              NaN                           19.875   
19970106                              NaN                           19.875   
19970107                              NaN                           19.875   
19970108                              NaN                           19.750   
19970109                              NaN                           19.875   
19970110                              NaN                           19.750   
19970113                              NaN                           19.875   
19970114                              NaN                           19.750   
19970115                              NaN                           19.750   
1

In [8]:
data_df_clean.shape

(5046, 21537)

We can observe that after cleaning we obtain a dataset consisting of 21537 companies.

In [9]:
data_df_clean.to_csv('new_sample.csv', sep=',')